### PR TITLE
Make HPACKHeaders inlinable.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

HPACKHeaders is a trivial data structure (a wrapper around Array) that
is present in the public API of all NIO HTTP/2 programs. This makes it a
tempting target for adding inlinable calls, primarily as a way to drive
down ARC. The fact that strings frequently travel across the API
boundary means that otherwise innocuous use-cases can trigger unexpected
ARC traffic.

Making as much of the implementation inlinable as possible reduces a lot
of this overhead and opens up the possibility for better performance in
a number of parts of the code, including in the core HTTP/2 protocol
engine, which operates extensively on this data type.

Modifications:

- Make basically everything inlinable.

Result:

Our benchmarks are about 0.7% faster.

### CODE REVIEW

File: Sources/NIOHPACK/HPACKHeader.swift [hunk 1]
```diff
@@ -486,14 +528,16 @@ extension String.UTF8View {
             }
     }
 
+    @inlinable
     @inline(never)
-    private func _compareCaseInsensitiveASCIIBytesSlowPath(to other: String.UTF8View) -> Bool {
+    func _compareCaseInsensitiveASCIIBytesSlowPath(to other: String.UTF8View) -> Bool {
```
Comment (PeterAdams-A) [inline Sources/NIOHPACK/HPACKHeader.swift:533]: This looks odd - inalienable but never?

File: Sources/NIOHPACK/HPACKHeader.swift [hunk 2]
```diff
@@ -461,7 +502,8 @@ extension String.UTF8View {
     ///
     /// - Parameter bytes: The string constant in the form of a collection of `UInt8`
     /// - Returns: Whether the collection contains **EXACTLY** this array or no, but by ignoring case.
-    fileprivate func compareCaseInsensitiveASCIIBytes(to other: String.UTF8View) -> Bool {
+    @inlinable
```
Comment (PeterAdams-A) [inline Sources/NIOHPACK/HPACKHeader.swift:505]: I've not read the rules but I don't see any logical reason why something shouldn't be both inlineable and fileprivate.

**Review by PeterAdams-A — APPROVED** [review]
Generally looks good but I've made a couple of comments.

File: Sources/NIOHPACK/HPACKHeader.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIOHPACK/HPACKHeader.swift:533]: Yup. This is a logically sensible stanza. `@inlinable` means the definition is available to the compiler when compiling the module using the code. This allows, for example, specialization. But the method may not be _inlined_. In this case the primary use is to let the compiler see through the call and observe that we never write to any memory here.

File: Sources/NIOHPACK/HPACKHeader.swift (hunk 2 above)
Comment (Lukasa) [reply Sources/NIOHPACK/HPACKHeader.swift:505]: Totally reasonable except that the rule is it has to be `internal` or higher.

**Review by glbrntt — APPROVED** [review]

**Pull request merged**